src/state/tls.py

```
@@ -76,31 +83,40 @@ def from_charm(
             )
 
         return cls(
-            external_hostname=external_hostname,
+            hostnames=hostnames,
             tls_cert_and_ca_chain=tls_cert_and_ca_chain,
             private_key=private_key,
         )
 
+    # Validation is done in this method instead of using a pydantic model because
+    # there are cases where we need to validate the state but we don't need the state instance.
     @classmethod
-    def validate(cls, charm: ops.CharmBase) -> None:
+    def validate(cls, charm: ops.CharmBase, certificates: TLSCertificatesRequiresV4) -> None:
         """Validate the precondition to initialize this state component.
 
         Args:
             charm: The haproxy charm.
+            certificates: TLS certificates requirer class.
 
         Raises:
             TLSNotReadyError: if the charm is not ready to handle TLS.
         """
-        tls_requirer_integration = charm.model.get_relation(TLS_CERTIFICATES_INTEGRATION)
-        external_hostname = typing.cast(str, charm.config.get("external-hostname", ""))
-
-        if not re.match(HOSTNAME_REGEX, external_hostname):
+        tls_requirer_integration = charm.model.get_relation(certificates.relationship_name)
+        if not certificates.certificate_requests:
+            logger.error("The charm did not request any certificates.")
+            raise TLSNotReadyError("The charm did not request any certificates.")
```

I may be lost in the PR: is it actually an error? Couldn't we have a ha-proxy with just http sites? Or are we already in a case here where we know we have some TLS but we are missing info (and the message may be slightly different in that case).

---

> Couldn't we have a ha-proxy with just http sites?
No, for ingress and haproxy-route we enfore TLS which means the charm should block if it didn't raise any cert request ( either `external-hostname` is not configured or requirer hasn't sent over any hostname ).

As we don't have any cert request it's up to the operator to either configure `external-hostname` or to relate haproxy to a requirer that requests a hostname ( that's why the status is blocked here rather than a waiting status since the responsibility falls on the operator to take action )